bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,132 @@
+        creds.values()
+            .all(|cred| cred.validate_internode())
+    }
+
     pub fn set_nodes_credentials(
         &mut self,
-        nodes: HashMap<IpAddr, Vec<Credentials>>,
+        mut nodes: HashMap<String, DeclaredCredentials>,
     ) -> Result<(), Error> {
-        if nodes
-            .values()
-            .all(|creds|
-                creds
-                    .iter()
-                    .all(|cred|
-                        cred.ip().is_some() &&
-                        cred.kind().map(|k| k.is_internode()) == Some(true)))
-        {
-            self.nodes = nodes;
+        if Self::node_creds_ok(&nodes) {
+            let mut nodes_creds = self.nodes.write().expect("nodes credentials lock");
+            for (nodename, cred) in nodes.drain() {
+                let guard = DCredentialsResolveGuard::new(cred.clone());
+                nodes_creds.insert(nodename, guard);
+                if cred.ip().is_empty() && cred.hostname().is_some() {
+                    self.spawn_resolver(cred);
+                }
+            }
             Ok(())
         } else {
             let message = "nodes credentials missing ip or node name";
             Err(Error::CredentialsNotProvided(message.to_string()))
         }
     }
 
-    fn check_node_request(&self, node_name: &String, ip: Option<IpAddr>) -> Option<bool> {
-        if self.nodes.is_empty() {
-            warn!("nodes credentials not set");
+    fn spawn_resolver(&self, cred: DeclaredCredentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_dur_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, cred: DeclaredCredentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = lookup(hostname).await;
+        let mut cur_sleep_dur_ms = 100;
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(cur_sleep_dur_ms)).await;
+
+            addr = lookup(hostname).await;
+
+            cur_sleep_dur_ms = sleep_dur_ms.min(cur_sleep_dur_ms * 2);
+        }
+
+        let addr = addr.expect("somehow addr is none");
+        if let CredentialsKind::InterNode(nodename) = cred.kind() {
+            let mut nodes = nodes.write().expect("nodes credentials lock");
+            if let Some(creds) = nodes.get_mut(nodename) {
+                creds.creds_mut().replace_addresses(addr);
+                creds.set_resolved();
+            }
+        } else {
+            error!("resolved credentials are not internode");
+        }
+    }
+
+    fn process_auth_result(&self, nodename: &str, unresolved: bool) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        if let Some(node) = nodes.get_mut(nodename) {
+            if node.update_resolve_state(unresolved) {
```

Comment:
With intensive access from a malicious node, just counting the consecutive number of failed authorizations is not enough to prevent DDoS. We should additionally check that the time interval passed from the previous successful resolution is larger than `resolve_sleep_dur_ms`. Otherwise, the malicious node will trigger re-resolving indefinitely without any pause